provider/aws_sd.go

```diff
@@ -0,0 +1,626 @@
+		dryRun:              dryRun,
+		ownerID:             ownerID,
+	}
+
+	return provider, nil
+}
+
+// Records returns list of all endpoints.
+func (p *AWSSDProvider) Records() (endpoints []*endpoint.Endpoint, err error) {
+	namespaces, err := p.ListNamespaces()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, ns := range namespaces {
+		services, err := p.ListServicesByNamespaceID(ns.Id)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, srv := range services {
+			instances, err := p.ListInstancesByServiceID(srv.Id)
+			if err != nil {
+				return nil, err
+			}
+
+			if len(instances) > 0 {
+				// DNS name of the record is a concatenation of service and namespace
+				dnsName := *srv.Name + "." + *ns.Name
+				ep, err := p.instancesToEndpoint(dnsName, srv, instances)
+				if err != nil {
+					return nil, err
+				}
+				endpoints = append(endpoints, ep)
+			}
+		}
+	}
+
+	return endpoints, nil
+}
+
+func (p *AWSSDProvider) instancesToEndpoint(recordName string, srv *sd.Service, instances []*sd.InstanceSummary) (*endpoint.Endpoint, error) {
+	newEndpoint := &endpoint.Endpoint{
+		DNSName:   recordName,
+		RecordTTL: endpoint.TTL(aws.Int64Value(srv.DnsConfig.DnsRecords[0].TTL)),
+		Targets:   make(endpoint.Targets, 0, len(instances)),
+	}
+
+	for _, inst := range instances {
+		// CNAME
+		if inst.Attributes["AWS_INSTANCE_CNAME"] != nil && aws.StringValue(srv.DnsConfig.DnsRecords[0].Type) == sd.RecordTypeCname {
+			newEndpoint.RecordType = endpoint.RecordTypeCNAME
+			newEndpoint.Targets = append(newEndpoint.Targets, aws.StringValue(inst.Attributes["AWS_INSTANCE_CNAME"]))
+
+			// ALIAS
+		} else if inst.Attributes["AWS_ALIAS_DNS_NAME"] != nil {
+			newEndpoint.RecordType = endpoint.RecordTypeCNAME
+			newEndpoint.Targets = append(newEndpoint.Targets, aws.StringValue(inst.Attributes["AWS_ALIAS_DNS_NAME"]))
+
+			// IP-based target
+		} else if inst.Attributes["AWS_INSTANCE_IPV4"] != nil {
+			newEndpoint.RecordType = endpoint.RecordTypeA
+			newEndpoint.Targets = append(newEndpoint.Targets, aws.StringValue(inst.Attributes["AWS_INSTANCE_IPV4"]))
+		} else {
+			return nil, fmt.Errorf("FATAL ERROR: Instance not valid \"%v\"", inst)
+		}
+	}
+
+	return newEndpoint, nil
+}
+
+// ApplyChanges applies Kubernetes changes in endpoints to AWS API
+func (p *AWSSDProvider) ApplyChanges(changes *plan.Changes) error {
+	// return early if there is nothing to change
+	if len(changes.Create) == 0 && len(changes.Delete) == 0 && len(changes.UpdateNew) == 0 {
+		log.Info("All records are already up to date")
+		return nil
+	}
+
+	// convert updates to delete and create operation if applicable (updates not supported)
+	creates, deletes := p.updatesToCreates(changes)
+	changes.Delete = append(changes.Delete, deletes...)
+	changes.Create = append(changes.Create, creates...)
+
+	namespaces, err := p.ListNamespaces()
+	if err != nil {
+		return err
+	}
+
+	err = p.submitDeletes(namespaces, changes.Delete)
+	if err != nil {
```

Comment:
if the external-dns crashes after this line execution, then we end up in state when records  to be updated are deleted, but not recreated ?

Reply:
It is a valid point 👍 But it must be in this order. There are two reasons:

* Update support - when just list of targets is updated `[1.2.3.4] -> [1.2.3.4, 1.2.3.5]` it is translated to:
```
deletes = [1.2.3.4]
creates = [1.2.3.4, 1.2.3.5]
```
Then when `deletes` are processed after `creates` it will miss the `1.2.3.4` instance.

* Executing creates first would double required resources. During the transition resource limit can be hit (100 instances per service or 2000 instances per namespace).
https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/DNSLimitations.html#limits-api-entities-autonaming